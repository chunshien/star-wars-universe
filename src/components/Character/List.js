import React, { Fragment, useCallback } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components'

import ListItem from './ListItem';
import Pagination from '../Pagination/Pagination';

const Content = styled.div`
    width: 100%;
    padding: 1%;
    float: left;
    box-sizing: border-box;
`;

const EmptyData = styled.div`
    padding: 10px;
    background-color: white;
    border-radius: 5px;
    box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2), 0px 4px 5px 0px rgba(0,0,0,0.14), 0px 1px 10px 0px rgba(0,0,0,0.12);
    text-align: center;
`;

function List(props) {
    const handleClick = useCallback((url) => {
        if(props){
            props.onClick(url)
        }
    }, [props])


    const renderList = () => {
        const {peoples, loading} = props;
        if (peoples && peoples.length > 0) {
            return peoples.map(item=>{
                return (
                    <ListItem key={item.url} data={item} onClick={handleClick} />
                )
            });
        } else {
            return (
                <div>
                    {!loading &&
                        <EmptyData>No data available</EmptyData>
                    }
                </div>
            )
        }
    }

    const handlePageChanged = useCallback((index) => {
        if(props){
            props.onPageChanged(index)
        }
    }, [props])

    const renderPagination = () => {
        const {peoples, totalItem, activePage} = props;
        if (totalItem > 0) {
            return (
                <Pagination 
                    totalItem={totalItem}
                    itemCountPerPage={peoples.length}
                    activePage={activePage}
                    onChanged={handlePageChanged}
                />
            )
        }
    }

    return (
        <Fragment>
            <Content>
                {renderList()}
            </Content>
            <Content>
                {renderPagination()}
            </Content>
        </Fragment>
    )
}
List.propTypes = {
    peoples: PropTypes.array,
    totalItem: PropTypes.number,
    activePage: PropTypes.number,
    itemCountPerPage: PropTypes.number,
    onChanged: PropTypes.func
};
export default List;