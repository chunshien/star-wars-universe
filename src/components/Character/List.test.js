import React from 'react';
import renderer from 'react-test-renderer';
import TestUtils from 'react-dom/test-utils';

import List from './List';
const testData = require('../../assets/test-data/peoples')

describe('Component - List Test', () => {
    it('renders List without data without crashing', () => {
        const component = renderer.create(
            <List/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders List with data without crashing', () => {
        const component = renderer.create(
            <List
                peoples={testData.results}
                loading={false}
                totalItem={testData.count}
                activePage={1}            
            />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    //mock props function call
    let ListElement;
    let onClickMock = jest.fn();
    let onPageChangedMock = jest.fn();
    class Wrapper extends React.Component {
        render() { 
            return this.props.children
        }
    }

    beforeAll(function() {
        ListElement = TestUtils.renderIntoDocument(
            <Wrapper>
                <List 
                    onClick={onClickMock}
                    onPageChanged={onPageChangedMock}
                />
            </Wrapper>
        );
    });
    beforeEach(function() {
        onClickMock.mockClear();
        onPageChangedMock.mockClear();
    });
    it('Simulate List onClick', () => {
        ListElement.props.children.props.onClick();
        expect(onClickMock).toBeCalled();                
    });

    it('Simulate ProductList onHandlePageChanged', () => {
        ListElement.props.children.props.onPageChanged();
        expect(onPageChangedMock).toBeCalled();     
    });
})