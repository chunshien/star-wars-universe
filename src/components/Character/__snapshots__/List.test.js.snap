// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - List Test renders List with data without crashing 1`] = `
Array [
  <div
    className="sc-AxheI fmZbrR"
  >
    <div
      className="sc-AxjAm iFUVjd"
      onClick={[Function]}
    >
      <div
        className="sc-AxirZ ehqiCf"
      >
        <div
          className="sc-AxiKw iPnqiR"
        >
          <img
            alt="Luke Skywalker"
            className="sc-AxhCb jvfHjD avatar"
            src="avatar.png"
            style={
              Object {
                "backgroundColor": "#ffe0bd",
              }
            }
          />
        </div>
        <div
          className="sc-AxiKw iPnqiR"
        >
          <div>
            Name: 
            Luke Skywalker
          </div>
          <div>
            Height: 
            172
          </div>
          <div>
            Gender: 
            male
          </div>
          <div>
            Mass: 
            77
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-AxjAm iFUVjd"
      onClick={[Function]}
    >
      <div
        className="sc-AxirZ ehqiCf"
      >
        <div
          className="sc-AxiKw iPnqiR"
        >
          <img
            alt="C-3PO"
            className="sc-AxhCb jvfHjD avatar"
            src="avatar.png"
            style={
              Object {
                "backgroundColor": "gold",
              }
            }
          />
        </div>
        <div
          className="sc-AxiKw iPnqiR"
        >
          <div>
            Name: 
            C-3PO
          </div>
          <div>
            Height: 
            167
          </div>
          <div>
            Gender: 
            n/a
          </div>
          <div>
            Mass: 
            75
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-AxjAm iFUVjd"
      onClick={[Function]}
    >
      <div
        className="sc-AxirZ ehqiCf"
      >
        <div
          className="sc-AxiKw iPnqiR"
        >
          <img
            alt="R2-D2"
            className="sc-AxhCb jvfHjD avatar"
            src="avatar.png"
            style={
              Object {
                "backgroundColor": "white, blue",
              }
            }
          />
        </div>
        <div
          className="sc-AxiKw iPnqiR"
        >
          <div>
            Name: 
            R2-D2
          </div>
          <div>
            Height: 
            96
          </div>
          <div>
            Gender: 
            n/a
          </div>
          <div>
            Mass: 
            32
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-AxjAm iFUVjd"
      onClick={[Function]}
    >
      <div
        className="sc-AxirZ ehqiCf"
      >
        <div
          className="sc-AxiKw iPnqiR"
        >
          <img
            alt="Darth Vader"
            className="sc-AxhCb jvfHjD avatar"
            src="avatar.png"
            style={
              Object {
                "backgroundColor": "white",
              }
            }
          />
        </div>
        <div
          className="sc-AxiKw iPnqiR"
        >
          <div>
            Name: 
            Darth Vader
          </div>
          <div>
            Height: 
            202
          </div>
          <div>
            Gender: 
            male
          </div>
          <div>
            Mass: 
            136
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-AxjAm iFUVjd"
      onClick={[Function]}
    >
      <div
        className="sc-AxirZ ehqiCf"
      >
        <div
          className="sc-AxiKw iPnqiR"
        >
          <img
            alt="Leia Organa"
            className="sc-AxhCb jvfHjD avatar"
            src="avatar.png"
            style={
              Object {
                "backgroundColor": "#ffd6a4",
              }
            }
          />
        </div>
        <div
          className="sc-AxiKw iPnqiR"
        >
          <div>
            Name: 
            Leia Organa
          </div>
          <div>
            Height: 
            150
          </div>
          <div>
            Gender: 
            female
          </div>
          <div>
            Mass: 
            49
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-AxjAm iFUVjd"
      onClick={[Function]}
    >
      <div
        className="sc-AxirZ ehqiCf"
      >
        <div
          className="sc-AxiKw iPnqiR"
        >
          <img
            alt="Owen Lars"
            className="sc-AxhCb jvfHjD avatar"
            src="avatar.png"
            style={
              Object {
                "backgroundColor": "#ffd6a4",
              }
            }
          />
        </div>
        <div
          className="sc-AxiKw iPnqiR"
        >
          <div>
            Name: 
            Owen Lars
          </div>
          <div>
            Height: 
            178
          </div>
          <div>
            Gender: 
            male
          </div>
          <div>
            Mass: 
            120
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-AxjAm iFUVjd"
      onClick={[Function]}
    >
      <div
        className="sc-AxirZ ehqiCf"
      >
        <div
          className="sc-AxiKw iPnqiR"
        >
          <img
            alt="Beru Whitesun lars"
            className="sc-AxhCb jvfHjD avatar"
            src="avatar.png"
            style={
              Object {
                "backgroundColor": "#ffd6a4",
              }
            }
          />
        </div>
        <div
          className="sc-AxiKw iPnqiR"
        >
          <div>
            Name: 
            Beru Whitesun lars
          </div>
          <div>
            Height: 
            165
          </div>
          <div>
            Gender: 
            female
          </div>
          <div>
            Mass: 
            75
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-AxjAm iFUVjd"
      onClick={[Function]}
    >
      <div
        className="sc-AxirZ ehqiCf"
      >
        <div
          className="sc-AxiKw iPnqiR"
        >
          <img
            alt="R5-D4"
            className="sc-AxhCb jvfHjD avatar"
            src="avatar.png"
            style={
              Object {
                "backgroundColor": "white, red",
              }
            }
          />
        </div>
        <div
          className="sc-AxiKw iPnqiR"
        >
          <div>
            Name: 
            R5-D4
          </div>
          <div>
            Height: 
            97
          </div>
          <div>
            Gender: 
            n/a
          </div>
          <div>
            Mass: 
            32
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-AxjAm iFUVjd"
      onClick={[Function]}
    >
      <div
        className="sc-AxirZ ehqiCf"
      >
        <div
          className="sc-AxiKw iPnqiR"
        >
          <img
            alt="Biggs Darklighter"
            className="sc-AxhCb jvfHjD avatar"
            src="avatar.png"
            style={
              Object {
                "backgroundColor": "#ffd6a4",
              }
            }
          />
        </div>
        <div
          className="sc-AxiKw iPnqiR"
        >
          <div>
            Name: 
            Biggs Darklighter
          </div>
          <div>
            Height: 
            183
          </div>
          <div>
            Gender: 
            male
          </div>
          <div>
            Mass: 
            84
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-AxjAm iFUVjd"
      onClick={[Function]}
    >
      <div
        className="sc-AxirZ ehqiCf"
      >
        <div
          className="sc-AxiKw iPnqiR"
        >
          <img
            alt="Obi-Wan Kenobi"
            className="sc-AxhCb jvfHjD avatar"
            src="avatar.png"
            style={
              Object {
                "backgroundColor": "#ffe0bd",
              }
            }
          />
        </div>
        <div
          className="sc-AxiKw iPnqiR"
        >
          <div>
            Name: 
            Obi-Wan Kenobi
          </div>
          <div>
            Height: 
            182
          </div>
          <div>
            Gender: 
            male
          </div>
          <div>
            Mass: 
            77
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="sc-AxheI fmZbrR"
  >
    <div
      className="sc-AxhUy fIZDaY"
    >
      <span
        className="sc-AxgMl djxOiP"
        onClick={[Function]}
        style={
          Object {
            "color": "#dddddd",
            "cursor": "not-allowed",
          }
        }
      >
        &lt; Previous page
      </span>
      <span
        className="sc-AxgMl djxOiP"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderBottom": "5px solid black",
          }
        }
      >
        1
      </span>
      <span
        className="sc-AxgMl djxOiP"
        onClick={[Function]}
        style={Object {}}
      >
        2
      </span>
      <span
        className="sc-AxgMl djxOiP"
        onClick={[Function]}
        style={Object {}}
      >
        3
      </span>
      <span
        className="sc-AxgMl djxOiP"
        onClick={[Function]}
        style={
          Object {
            "color": "black",
          }
        }
      >
        Next page &gt;
      </span>
    </div>
  </div>,
]
`;

exports[`Component - List Test renders List without data without crashing 1`] = `
Array [
  <div
    className="sc-AxheI fmZbrR"
  >
    <div>
      <div
        className="sc-Axmtr hRgGIe"
      >
        No data available
      </div>
    </div>
  </div>,
  <div
    className="sc-AxheI fmZbrR"
  />,
]
`;
